### PR TITLE
Change how prep period is set

### PR DESCRIPTION
**Pull request by Sybrand**
*Overview:* by requiring start and end dates on all requests, we gain clarity, get rid of a bunch of times we allow null, and we get rid of the "set prep period" call, by just assuming we have to create a prep period if one doesn't exist.
*Side effect:* if you change the prep period, you "lose" all the changes you've made, since a new prep period record is created.
*Side effect:* this already existed, but even more so now - changing a fire start on a given day doesn't mean that changes for all other prep periods - it applies ONLY to the prep period you're working on.

- [x] By default, load the most recently edited prep period that overlaps with the current date.
- [x] Only save a prep period in the database when some value is changed.
- [x] Modify warning message.
- [x] Update tests
# Test Links:
[Percentile Calculator](https://wps-pr-1877.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-1877.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-1877.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-1877.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-1877.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=1108&f=c5&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN#state=2ec784ca-c46a-49d0-b2b3-1cf32a9015a2&session_state=7d9447c8-db66-4661-b4cb-03d2ac0d1d8f&code=32292df4-2bdf-4f90-a4a8-c8dbcda682a9.7d9447c8-db66-4661-b4cb-03d2ac0d1d8f.2b63f390-f3dc-43ae-89f2-016453863476)
[Fire Behaviour Advisory](https://wps-pr-1877.apps.silver.devops.gov.bc.ca/fire-behaviour-advisory)
[HFI Calculator](https://wps-pr-1877.apps.silver.devops.gov.bc.ca/hfi-calculator)
[FWI Calculator](https://wps-pr-1877.apps.silver.devops.gov.bc.ca/fwi-calculator)

### CODE REVIEW

File: api/app/routers/hfi_calc.py [hunk 1]
```diff
@@ -40,18 +42,21 @@
...
-                                                         fire_centre_id,
-                                                         date_range.start_date)
+    if date_range:
+        stored_request = get_most_recent_updated_hfi_request(session,
+                                                             fire_centre_id,
+                                                             date_range)
+        # NOTE: We could be real nice here, and look for a prep period that intercepts, and grab data there.
+    else:
+        # No date range specified!
+        stored_request = get_more_recent_updated_hfi_request_in_range(session, fire_centre_id)
```
Comment (conbrad) [inline api/app/routers/hfi_calc.py:59]: Shouldn't this be the other way around? If there's no date range how would we look up a request in a range?

**Review by conbrad — CHANGES_REQUESTED** [review]

File: api/app/routers/hfi_calc.py (hunk 1 above)
Comment (Sybrand) [reply api/app/routers/hfi_calc.py]: Bad function name - I've renamed it `get_more_recent_updated_hfi_request_for_current_date` ; So it just looks at the server date to figure out the relevant request.

**Review by conbrad — APPROVED** [review]

File: api/app/db/crud/hfi_calc.py [hunk 2]
```diff
@@ -35,15 +35,23 @@ def get_fire_centre_stations(session, fire_centre_id: int) -> CursorResult:
...
-        query = query.filter(HFIRequest.prep_start_day == prep_start_day)
-    if prep_end_day is not None:
-        query = query.filter(HFIRequest.prep_end_day == prep_end_day)
+        .filter(HFIRequest.fire_centre_id == fire_centre_id)\
+        .filter(HFIRequest.prep_start_day == date_range.start_date)\
+        .filter(HFIRequest.prep_end_day == date_range.end_date)
+    return query.order_by(HFIRequest.create_timestamp.desc()).first()
+
+
+def get_more_recent_updated_hfi_request_for_current_date(session: Session,
```
Comment (andrea-williams) [inline api/app/db/crud/hfi_calc.py:47]: Wouldn't `get_most_recent_updated_...` be a more accurate/descriptive name for this function?

**Review by andrea-williams — APPROVED** [review]

**Review by tessmgood — APPROVED** [review]

**Review by conbrad — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,10 @@
+## 2022-04-11 HFI Calc [#1634](https://github.com/bcgov/wps/issues/1634)
+
+### Features
+
+- **hfi calculator:** Load the most recently saved prep record that overlaps with the current date. (Previous behaviour was to load the most recently modified prep record.)
+
+- **hfi calculator:** Changing the prep period date does not result in a write to the database. Users can change thus change dates without affecting other users. (Previous behaviour was to create/update a prep record when changing the date.)
+- **hfi calculator:** When a fire starts are changed or weather stations are toggled, a prep period record is created in the database.
+- **hfi calculator:** Changing the prep period end date results in losing fire starts & weather station toggles. (Previous behaviour would retain fire starts & weather station toggles if the start start date was unchanged.)
+- **hfi calculator:** Changed warning message from "Any changes made to dates, fire starts or selected stations ..." to "Any changes made to fire starts or selected stations..."
```
Comment (andrea-williams) [inline CHANGELOG.md:10]: If I'm being REALLLLLLY picky/conscientious about finding a good balance between enough/too much detail for our PO or users, I'd say that this last bullet is prob too trivial a change for users to NEED to know about.

File: CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+## 2022-04-11 HFI Calc [#1634](https://github.com/bcgov/wps/issues/1634)
+
+### Features
+
+- **hfi calculator:** Load the most recently saved prep record that overlaps with the current date. (Previous behaviour was to load the most recently modified prep record.)
+
+- **hfi calculator:** Changing the prep period date does not result in a write to the database. Users can change thus change dates without affecting other users. (Previous behaviour was to create/update a prep record when changing the date.)
+- **hfi calculator:** When a fire starts are changed or weather stations are toggled, a prep period record is created in the database.
+- **hfi calculator:** Changing the prep period end date results in losing fire starts & weather station toggles. (Previous behaviour would retain fire starts & weather station toggles if the start start date was unchanged.)
```
Comment (andrea-williams) [inline CHANGELOG.md:9]: "start start" repeated repeated

**Review by andrea-williams — APPROVED** [review]

**Pull request merged**